File: src/bumperbot_examples/src/bumperbot_examples/tf_examples.py
# !/usr/bin/env python3
import rospy
from tf2_ros import StaticTransformBroadcaster, TransformBroadcaster, TransformListener, Buffer
from geometry_msgs.msg import TransformStamped
from bumperbot_examples.srv import GetTransform, GetTransformResponse

class TfExamples(object):
    def __init__(self):
        self.static_broadcaster_ = StaticTransformBroadcaster()
        self.dynamic_broadcaster_ = TransformBroadcaster()
        self.static_transform_stamped_ = TransformStamped()
        self.dynamic_transform_stamped_ = TransformStamped()

        self.timer_ = rospy.Timer(rospy.Duration(0.1), self.timerCallback)

        self.x_increment_ = 0.05
        self.last_x_ = 0.0

        self.static_transform_stamped_.header.stamp = rospy.Time.now()
        self.static_transform_stamped_.header.frame_id = "bumperbot_base"
        self.static_transform_stamped_.child_frame_id = "bumperbot_top"

        self.static_transform_stamped_.transform.translation.x = 0.0
        self.static_transform_stamped_.transform.translation.y = 0.0
        self.static_transform_stamped_.transform.translation.z = 0.3
        # quaternion
        self.static_transform_stamped_.transform.rotation.x = 0
        self.static_transform_stamped_.transform.rotation.y = 0
        self.static_transform_stamped_.transform.rotation.z = 0
        self.static_transform_stamped_.transform.rotation.w = 1

        self.static_broadcaster_.sendTransform(self.static_transform_stamped_)
        rospy.loginfo("Publishing static transform between %s and %s", self.static_transform_stamped_.header.frame_id, 
                      self.static_transform_stamped_.child_frame_id)
        
        self.get_transform_srv_ = rospy.Service("get_transform", GetTransform, self.getTransformCallback)
        self.tf_buffer_ = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer_)
        
    def timerCallback(self, event):
        self.dynamic_transform_stamped_.header.stamp = rospy.Time.now()
        self.dynamic_transform_stamped_.header.frame_id = "odom"
        self.dynamic_transform_stamped_.child_frame_id = "bumperbot_base"

        self.dynamic_transform_stamped_.transform.translation.x = self.last_x_ + self.x_increment_
        self.dynamic_transform_stamped_.transform.translation.y = 0
        self.dynamic_transform_stamped_.transform.translation.z = 0

        self.dynamic_transform_stamped_.transform.rotation.x = 0
        self.dynamic_transform_stamped_.transform.rotation.y = 0
        self.dynamic_transform_stamped_.transform.rotation.z = 0
        self.dynamic_transform_stamped_.transform.rotation.w = 1

        self.dynamic_broadcaster_.sendTransform(self.dynamic_transform_stamped_)
        self.last_x_ = self.dynamic_transform_stamped_.transform.translation.x

    def getTransformCallback(self, req):
        rospy.loginfo("Requested transform between %s and %s", req.frame_id, req.child_frame_id)
        res = GetTransformResponse()
        requested_transform = TransformStamped()

        try:
            requested_transform = self.tf_buffer_.lookup_transform(req.frame_id, req.child_frame_id, rospy.Time())
        except Exception as e:
            rospy.logerr("An error occured while transforming %s and %s", req.frame_id, req.child_frame_id)
            res.success = False
            return res

        rospy.loginfo("The requested transform is: %s", requested_transform)
        res.transform = requested_transform
        res.success = True
        return res

File: src/bumperbot_examples/src/tf_examples.cpp
// http://docs.ros.org/en/melodic/api/geometry_msgs/html/msg/TransformStamped.html
// http://wiki.ros.org/tf2_ros?distro=noetic
// https://wiki.ros.org/tf/Tutorials/Writing%20a%20tf%20broadcaster%20(C++)
// http://wiki.ros.org/tf2/Tutorials/Writing%20a%20tf2%20listener%20%28C%2B%2B%29

#include "bumperbot_examples/tf_examples.h"
#include <tf2_ros/static_transform_broadcaster.h>
#include <tf2_ros/transform_broadcaster.h>

TfExamples::TfExamples(const ros::NodeHandle &nh) 
    : nh_(nh), last_x_(0.0), x_increment_(0.05), tf_listener_(tf_buffer_) {
    static tf2_ros::StaticTransformBroadcaster static_broadcaster;

    static_transform_stamped_.header.stamp = ros::Time::now();
    static_transform_stamped_.header.frame_id = "bumperbot_base";
    static_transform_stamped_.child_frame_id = "bumperbot_top";

    static_transform_stamped_.transform.translation.x = 0.0;
    static_transform_stamped_.transform.translation.y = 0.0;
    static_transform_stamped_.transform.translation.z = 0.3;

    static_transform_stamped_.transform.rotation.x = 0;
    static_transform_stamped_.transform.rotation.y = 0;
    static_transform_stamped_.transform.rotation.z = 0;
    static_transform_stamped_.transform.rotation.w = 1;

    static_broadcaster.sendTransform(static_transform_stamped_);
    ROS_INFO_STREAM("Publishing static transform between " << static_transform_stamped_.header.frame_id <<
                    " and " << static_transform_stamped_.child_frame_id);

    timer_ = nh_.createTimer(ros::Duration(0.1), &TfExamples::timerCallback, this);

    get_transform_srv_ = nh_.advertiseService("get_transform", &TfExamples::getTransformCallback, this);
}

void TfExamples::timerCallback(const ros::TimerEvent &event){
    static tf2_ros::TransformBroadcaster dynamic_broadcaster;

    dynamic_transform_stamped_.header.stamp = ros::Time::now();
    dynamic_transform_stamped_.header.frame_id = "odom";
    dynamic_transform_stamped_.child_frame_id = "bumperbot_base";

    dynamic_transform_stamped_.transform.translation.x = last_x_ + x_increment_;
    dynamic_transform_stamped_.transform.translation.y = 0.0;
    dynamic_transform_stamped_.transform.translation.z = 0.0;

    dynamic_transform_stamped_.transform.rotation.x = 0.0;
    dynamic_transform_stamped_.transform.rotation.y = 0.0;
    dynamic_transform_stamped_.transform.rotation.z = 0.0;
    dynamic_transform_stamped_.transform.rotation.w = 1.0;

    dynamic_broadcaster.sendTransform(dynamic_transform_stamped_);
    last_x_ = dynamic_transform_stamped_.transform.translation.x;
}

bool TfExamples::getTransformCallback(
    bumperbot_examples::GetTransform::Request &req,
    bumperbot_examples::GetTransform::Response &res){
    ROS_INFO_STREAM("Requested transform between " << req.frame_id <<  " and " << req.child_frame_id);

    geometry_msgs::TransformStamped requested_transform;

    try{
        requested_transform = tf_buffer_.lookupTransform(req.frame_id, req.child_frame_id, ros::Time(0));
    }
    catch(const tf2::TransformException &ex){
        ROS_ERROR_STREAM("An error occurred while transforming " << req.frame_id << " and " << req.child_frame_id << ": " << ex.what());
        res.success = false;
        return true;
    }

    ROS_INFO_STREAM("The requested transform is:\n" << requested_transform);
    res.transform = requested_transform;
    res.success = true;
    return true;
}